### PR TITLE
instance Prim SMGen

### PR DESCRIPTION
**Pull request by curiousleo**
Extracted from https://github.com/lehins/splitmix/commits/new-random.

This is mostly to get an idea for how open you would be to to this idea, @phadej.

For context: the new version of `random` we're working on over at https://github.com/idontgetoutmuch/random has a fast monadic PRNG implementation for all `RandomGen` instances that also implement `Prim`.

In addition, the new version of `random` will use this library's implementation of SplitMix as its default PRNG.

It would be nice for the default PRNG to be fast, so having a `Prim` instance for `SMGen` would be really good. Defining that instance requires looking into the data type itself, and the `SMGen` constructor is not exported, so the instance declaration must live in `splitmix` itself. That's what this PR adds.

Alternative: https://github.com/idontgetoutmuch/random/pull/76

### CODE REVIEW

**Comment by phadej** [comment]
I don't understand the motivation. How this instance would make PRNG to be fast? `splitmix` is fast.

---

I have to think hard about `primitive` dependency. I don't like an idea depending on it.

Also the documentation for `Prim` says 

> Class of types supporting primitive array operations. This includes interfacing with GC-managed memory (functions suffixed with ByteArray#) and interfacing with unmanaged memory (functions suffixed with Addr#). Endianness is platform-dependent.

I don't understand, why would you want an array of `SMGen`s?

Wouldn't something like `System.Random.SplitMix.Mutable` be more suitable, working in `PrimMonad`? If you need stuff to be fast?

**Comment by curiousleo** [comment]
> I don't understand the motivation. How this instance would make PRNG to be fast? `splitmix` is fast.

Yes, `splitmix` is fast. That's why we're keen to use it!

I didn't explain the motivation well, sorry. Let me try again. Here is an excerpt from the new module documentation:

```haskell
-- Given a /pure/ pseudo-random number generator, you can run it in an 'IO' or
-- 'ST' context using 'runPrimGenIO' or 'runPrimGenST' and their variants,
-- respectively. If the pseudo-random number generator does not implement
-- 'Prim', you can also use 'runMutGenIO' or 'runMutGenST' and their variants.
--
-- >>> let pureGen = mkStdGen 42
-- >>> runPrimGenIO_ pureGen (rolls 10) :: IO [Word8]
-- [1,1,3,2,4,5,3,4,6,2]
```
https://github.com/idontgetoutmuch/random/blob/module-level-docs/System/Random.hs#L158-L165

The new `random` library ships with a few monadic adapters that make it easier to use pure pseudo-random number generators like `SMGen` in a monadic (`PrimMonad`) context. The fastest of those _adapters_ is `PrimGen`, which requires the PRNG to have a `Prim` instance.
 
> I don't understand, why would you want an array of `SMGen`s?

In the new `random`, the `Prim` instance is used to quickly save and restore the state of the PRNG. So in this use case, we don't use an array of multiple `SMGen`s, we just save and restore the state of the PRNG into/from an array.

See https://github.com/idontgetoutmuch/random/blob/module-level-docs/System/Random.hs#L768-L772.

> Wouldn't something like `System.Random.SplitMix.Mutable` be more suitable, working in `PrimMonad`? If you need stuff to be fast?

Does `System.Random.SplitMix.Mutable` exist?
The idea of how the new `random` library is structured is to make it possible to use any pure PRNG as a monadic/mutable PRNG if the user wants to. The fastest way we have found to do this is via `Prim` instances in `PrimMonad` contexts.

I hope this makes sense. @lehins is definitely better placed to talk about the design decisions related to monadic adapters like `PrimGen`.

**Comment by curiousleo** [comment]
I'm aware that this is a chicken-and-egg problem: right now, there isn't much point in `SMGen` having a `Prim` instance because the new `random` is not yet released. But we don't really want to release `random` with less-than-optimal defaults.

One option would be to put the `Prim` instance and the `primitive` dependency behind a flag such that both are _not_ built/required by default.

**Comment by phadej** [comment]
> Does `System.Random.SplitMix.Mutable` exist?

No. But it could. I don't conceptually like having `Prim` instance when you really want a mutable variant of `SMGen` (and `freeze` + `thaw` to go between them). OTOH, exposing enough stuff from `splitmix` to make possible to make `splitmix-mutable` is fine.

I'm worried that `random` design is done too much performance in mind. E.g. `QuickCheck` won't benefit from `PrimGen` abstraction at all, or at least not without large rewriting. (And I don't think that it will even then, except for individual "leaf" generation of say random `ByteString`s).

**Comment by lehins** [comment]
@phadej the "new" random provides exactly the interface you are talking about:
```haskell
thawGen :: Frozen g -> m g
freezeGen :: g -> m (Frozen g)
```
Specifics of those type signatures aren't too important for this discussion, but it does provide a mutable interface not only for splitmix, but also for mwc-random and other mutable RNGs.

> I'm worried that random design is done too much performance in mind. E.g. QuickCheck won't benefit from PrimGen abstraction at all, or at least not without large rewriting.

No need to worry, mutable monadic interface is an addition, not the replacement for current interface, so QucikCheck will not need to be rewritten. That being said, much other work on performance improvements that has been put into `random` which will benefit QuickCheck (and others) greatly. We are talking about a factor of x50 faster generation of values.

I understand you reluctance for this PR and dependency on `primitive` does not bring anything to the table for `splitmix`. This is totally ok and we can include that instance into `random` directly as an orphan. It would be nice though if the `SMGen` constructor was exported, this way there would be no need to go through a `(Word64, Word64)` for creating `Prim` instance. Would you be OK with a PR that exports the constructor?

Side note. This was funny to read cause it was coming from a person that puts quite a bit of effort in performance. splitmix is a great proof of that ;)

>  too much performance in mind

**Comment by phadej** [comment]
> too much performance in mind

I didn't sacrifice any of interface elegance, did I?

**Comment by phadej** [comment]
I don't want to expose `SMGen` constructor as there is an invariant on the hidden field (it have to be odd). I'd rather write `System.Random.SplitMix.Random` with proper `SMGen s` type, even it's more work for me.

**Comment by lehins** [comment]
On the other hand, if you'd rather write `SMGen s` in a separate library `splitmix-mutable` or something that would work too.

**Comment by phadej** [comment]
I want specifically to mention, that Split in SplitMix is important. I hope that there will be good support for splittable random generators in `random`, also in mutable interface (it does make complete sense to have).

Also having `nextWord64` (and other operation variants) which would give thawed version of generator, so users don’t need to unsafeThaw.

**Comment by curiousleo** [comment]
Thanks all for the productive discussion!

We have for now removed the `primitive` dependency from the new `random` proposal: https://github.com/idontgetoutmuch/random/pull/80. This also makes this PR here unnecessary for the time being.

The idea is that at some later stage there will be a separate `random-primitive` package bringing back a fast monadic pseudo-random number generator adapter for pure pseudo-random number generators that implement `Prim` (something like the `PrimGen` that https://github.com/idontgetoutmuch/random/pull/80 removed).

> I want specifically to mention, that Split in SplitMix is important. I hope that there will be good support for splittable random generators in `random`, also in mutable interface (it does make complete sense to have).

@phadej good point! See https://github.com/idontgetoutmuch/random/pull/80/files#diff-4a522d098452ebfc9be2d6291588f5bcR540-R543.
